megazords/ios-rust/build-xcframework.sh

```diff
@@ -0,0 +1,173 @@
+REPO_ROOT="$( dirname "$( dirname "$THIS_DIR" )" )"
+
+MANIFEST_PATH="$THIS_DIR/Cargo.toml"
+if [[ ! -f "$MANIFEST_PATH" ]]; then
+  echo "Could not locate Cargo.toml relative to script"
+  exit 1
+fi
+
+CRATE_NAME=$(grep --max-count=1 '^name =' "$MANIFEST_PATH" | cut -d '"' -f 2)
+if [[ -z "$CRATE_NAME" ]]; then
+  echo "Could not determine crate name from $MANIFEST_PATH"
+  exit 1
+fi
+
+LIB_NAME="lib${CRATE_NAME}.a"
+
+####
+##
+## 1) Build the rust code individually for each target architecture.
+##
+####
+
+# Helper to run the cargo build command in a controlled environment.
+# It's important that we don't let environment variables from the user's default
+# desktop build environment leak into the iOS build, otherwise it might e.g.
+# link against the desktop build of NSS.
+
+CARGO="$HOME/.cargo/bin/cargo"
+LIBS_DIR="$REPO_ROOT/libs"
+
+DEFAULT_RUSTFLAGS=""
+BUILD_ARGS=(build --manifest-path "$MANIFEST_PATH" --lib)
+case $BUILD_PROFILE in
+  debug) ;;
+  release)
+    BUILD_ARGS=("${BUILD_ARGS[@]}" --release)
+    # With debuginfo, the zipped artifact quickly baloons to many
+    # hundred megabytes in size. Ideally we'd find a way to keep
+    # the debug info but in a separate artifact.
+    DEFAULT_RUSTFLAGS="-C debuginfo=0"
+    ;;
+  *) echo "Unknown build profile: $BUILD_PROFILE"; exit 1;
+esac
+
+cargo_build () {
+  TARGET=$1
+  case $TARGET in
+    x86_64*)
+      LIBS_DIR="$REPO_ROOT/libs/ios/x86_64";;
+    # TODO: when we want to include crates that depend on SQLCipher or NSS,
+    # we'll need to distinguish between hardware and simulator builds here
+    # and link the later against separately-compiled libraries.
+    # Ref https://github.com/mozilla/application-services/issues/4352.
+    aarch64*)
+      LIBS_DIR="$REPO_ROOT/libs/ios/arm64";;
+    *)
+      echo "Unexpected target architecture: $TARGET" && exit 1;;
+  esac
+  env -i \
+    NSS_STATIC=1 \
+    NSS_DIR="$LIBS_DIR/nss" \
+    SQLCIPHER_STATIC=1 \
+    SQLCIPHER_LIB_DIR="${LIBS_DIR}/sqlcipher/lib" \
+    SQLCIPHER_INCLUDE_DIR="${LIBS_DIR}/sqlcipher/include" \
+    PATH="${PATH}" \
+    RUSTC_WRAPPER="${RUSTC_WRAPPER:-}" \
+    SCCACHE_IDLE_TIMEOUT="${SCCACHE_IDLE_TIMEOUT:-}" \
+    SCCACHE_CACHE_SIZE="${SCCACHE_CACHE_SIZE:-}" \
+    SCCACHE_ERROR_LOG="${SCCACHE_ERROR_LOG:-}" \
+    RUST_LOG="${RUST_LOG:-}" \
+    RUSTFLAGS="${RUSTFLAGS:-$DEFAULT_RUSTFLAGS}" \
+    "$CARGO" "${BUILD_ARGS[@]}" --target "$TARGET"
+}
+
+set -euvx
+
+# Intel iOS simulator
+# TODO: why is the env var necessary?
+CFLAGS_x86_64_apple_ios="-target x86_64-apple-ios" \
+  cargo_build x86_64-apple-ios
+
+# Hardware iOS targets
+cargo_build aarch64-apple-ios
+
+# M1 iOS simulator.
+# It's currently in Nightly only and requires to build `libstd`.
+# We hope this will be available by default in Rust 1.56.0.
+BUILD_ARGS=(+nightly "${BUILD_ARGS[@]}" -Z build-std)
+cargo_build aarch64-apple-ios-sim
+
+# TODO: would it be useful to also include desktop builds here?
```

Comment:
🤔 Could be a good spike for a later time (to explore running tests via `swift test` like you mentioned in our chat yesterday)

Reply:
Right, exactly! I actually prototyped this a little bit and it seemed like it would work, but:

* Glean doesn't publish desktop builds, so we'd have to bring Glean into our megazord before we could do this
* It makes the compiled bundle even bigger, which may be a problem for production releases